File: leetcode/easy/Remove_Duplicates_Sorted_Array.java
/*
Type: Easy
26. Remove Duplicates from Sorted Array
https://leetcode.com/problems/remove-duplicates-from-sorted-array/description/?envType=study-plan-v2&envId=top-interview-150
*/

package easy;

import java.util.Hashtable;

public class Remove_Duplicates_Sorted_Array {
    static class Solution {
        public int removeDuplicates(int[] nums) {
            int j = 1;
            for (int i = 1; i < nums.length; i++) {
                if (nums[i] != nums[i - 1]) {
                    nums[j] = nums[i];
                    j++;
                }
            }
            return j;
        }
    }

//    This solution is not as efficient as the first one
    static class Solution2 {
        public int removeDuplicates(int[] nums) {
            Hashtable<Integer, Integer> num_hashtable = new Hashtable<>();
            int index = 0;
            for (int i = 0; i < nums.length; ++i) {
                if (num_hashtable.get(nums[i]) == null) {
                    num_hashtable.put(nums[i], i);
                    nums[index] = nums[i];
                    index++;
                }
            }
            return index;
        }
    }

    public static void main(String[] args) {
        int[] nums = new int[] {1,1,2};

        Solution solution = new Solution();
        System.out.println(solution.removeDuplicates(nums));
    }

}

File: leetcode/top_150_interview/2_medium/remove_duplicates_sorted_array.py
"""
Type: Medium
80. Remove Duplicates from Sorted Array II
https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/description/?envType=study-plan-v2&envId=top-interview-150
"""
from typing import List


class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        cur_number = None
        counter = 0
        len_nums = len(nums)
        index = 0
        while index < len_nums:
            if cur_number != nums[index]:
                cur_number = nums[index]
                counter = 1
            else:
                counter += 1
                if counter > 2:
                    nums.pop(index)
                    index -= 1
                    len_nums -= 1
            index += 1
